mm/map/mm_map.c

```diff
@@ -0,0 +1,325 @@
+
+FAR struct mm_map_entry_s *mm_map_next(
+                           FAR const struct mm_map_entry_s *entry)
+{
+  FAR struct tcb_s *tcb = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR struct mm_map_s *mm = &group->tg_mm_map;
+  FAR struct mm_map_entry_s *map = NULL;
+
+  if (mm_map_lock(group) == OK)
+    {
+      if (entry == NULL)
+        {
+          map = (struct mm_map_entry_s *)sq_peek(&mm->mm_map_sq);
+        }
+      else
+        {
+          map = (struct mm_map_entry_s *)sq_next(((sq_entry_t *)entry));
+        }
+
+      mm_map_unlock(group);
+    }
+
+  return map;
+}
+
+/****************************************************************************
+ * Name: mm_map_find
+ *
+ * Description:
+ *   Find the first mapping containing the range from the task group's list
+ *
+ ****************************************************************************/
+
+FAR struct mm_map_entry_s *mm_map_find(FAR const void *vaddr, size_t length)
+{
+  FAR struct tcb_s *tcb = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR struct mm_map_s *mm = &group->tg_mm_map;
+  FAR struct mm_map_entry_s *map = NULL;
+
+  if (mm_map_lock(group) == OK)
+    {
+      map = (struct mm_map_entry_s *)sq_peek(&mm->mm_map_sq);
+
+      while (map && !in_range(vaddr, length, map->vaddr, map->length))
+        {
+          map = (struct mm_map_entry_s *)sq_next(((sq_entry_t *)map));
+        }
+
+      mm_map_unlock(group);
+    }
+
+  return map;
+}
+
+/****************************************************************************
+ * Name: mm_map_remove
+ *
+ * Description:
+ *   Remove a mapping from the task  group's list
+ *
+ ****************************************************************************/
+
+int mm_map_remove(FAR struct task_group_s *group,
+                  FAR struct mm_map_entry_s **map)
+{
+  FAR struct mm_map_entry_s *prev;
+  FAR struct mm_map_entry_s *r = NULL;
+  FAR struct mm_map_s *mm;
+
+  if (!group)
+    {
+      /* We end up in here only through mm_map_destroy. The address
+       * environment is being destroyed and the group is not valid.
+       * just delete the entry
+       */
+
+      goto free_mem;
+    }
+
+  mm = &group->tg_mm_map;
+
+  int ret = mm_map_lock(group);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  prev = (struct mm_map_entry_s *)sq_peek(&mm->mm_map_sq);
+
+  /* Check if the list was empty */
+
+  if (!prev)
+    {
+      mm_map_unlock(group);
+      return -ENOENT;
+    }
+
+  /* Check if removing the first item */
+
+  if (*map == prev)
+    {
+      sq_remfirst(&mm->mm_map_sq);
+      *map = NULL;
+      r = prev;
+    }
+  else
+    {
+      /* Loop through the remaining items to find the one to be removed */
+
+      while ((r = (struct mm_map_entry_s *)sq_next(((sq_entry_t *)prev))))
+        {
+          if (*map == r)
+            {
+              sq_remafter((sq_entry_t *)prev, &mm->mm_map_sq);
+              *map = prev;
```

Comment:
Nowhere any more. I don't think that it has any use any more, this has been changed so many times. I'll just remove it

Reply:
done

Reply:
It is no point in adding group or map as the first argument for any functions in mm_map. You always work with currently running process' maps, and there is just no point in passing that as an argument. In fact, it is bad, because someone might think that he/she could start tinkering with some other process mappings, which is just never possible.

Only exception is the mapping removal, because then the context may be in fact invalid and shouldn't be used at all for anything. Thats why the group is being passed there so that the drivers can check if it is NULL. If it is not NULL, it is always the current of course. I agree that it is maybe not the cleanest solution, but for the mm_map_remove function there just has to be some way to pass the info that the context is invalid. Because that really happens at process destruction, and cleanup of kernel resources is still required.

Reply:
In fact, the group argument can be deleted also in mm_map_lock and mm_map_unlock, as also there you'll be only able to lock your own process' mappings. Certainly not any other's

Reply:
The current code mixes all possible combination:

1. Some pass task_group_s as the first argument
2. Some directly access the current task internally

It isn't good. The better method  is using only either method 1 or method 2 but not both. So, I suggest method 1 because it's more flexible. But if you prefer you can stick to the method 2.

Reply:
Added a patch to fix this if

Reply:
And yes, you are absolutely right. These two shouldn't be mixed. I want to stick to 2., but have an exception for mm_map_remove for the practical reason that process exit case cannot be handled without some extra argument indicating that the context is not valid